bleak/__init__.py

```diff
@@ -204,6 +205,37 @@ def set_scanning_filter(self, **kwargs):
         )
         self._backend.set_scanning_filter(**kwargs)
 
+    async def observe(self) -> AsyncIterator[Tuple[BLEDevice, AdvertisementData]]:
+        """
+        Yields devices and their advertising data as they are discovered.
+
+        .. note::
+            Ensure that scanning is started before calling this method.
+
+        Returns:
+            An async iterator that yields tuples (:class:`BLEDevice`, :class:`AdvertisementData`).
+        """
+        # noinspection PyProtectedMember
+        user_cb = self._backend._callback
+
+        devices = asyncio.Queue()
+
+        def detection_callback(bd, ad):
+            devices.put_nowait((bd, ad))
+            if user_cb:
+                user_cb(bd, ad)
+
+        self._backend.register_detection_callback(detection_callback)
+        try:
+            while True:
+                yield await devices.get()
+        finally:
+            # noinspection PyProtectedMember
+            if self._backend._callback not in (user_cb, detection_callback):
+                # This check can be removed when deprecated `register_detection_callback()` method is removed
+                raise BleakError("Detection callback has been changed while scanning")
+            self._backend.register_detection_callback(user_cb)
```

Comment:
This showcases why I did https://github.com/hbldh/bleak/pull/1366 - to avoid replicating the callback conversion logic as mentioned in https://github.com/hbldh/bleak/pull/1361#discussion_r1258874945 but still allow both functionalities to coexist (also the starting word https://github.com/hbldh/bleak/pull/1361/files#diff-48384236075c00b8fc59ec560483d0b5692082ebe327f27ac5f6aed55d73835eR72 is "_Additionally, ..._").

However, I don't really like this solution, and I can't really find a use case for using both methods concurrently. And yes, the check is done to handle if the user called `scanner.register_detection_callback()` while iterating over `scanner.observe()`, which they can.

Reply:
This is solved by #1367, see https://github.com/hbldh/bleak/pull/1361#discussion_r1268395747